WIDEFIELD: Spatial analysis

iezqrom opened the pull request.

iezqrom committed:
Add TraceMetrics integration to WidefieldAnalysis for trace analysis

- Updated __init__.pyi to include TraceMetrics in exports.
- Implemented get_trace_metrics method in WidefieldAnalysis for analyzing 1D traces with detailed documentation and error handling.

iezqrom committed:
Auto-format code

iezqrom committed:
Add motion correction functionality to WidefieldAnalysis

- Introduced methods for estimating and applying motion correction, including `estimate_motion_vectors` and `apply_motion_correction`.
- Added support for saving and loading motion vectors and reference images.
- Updated __init__.pyi to export new motion-related functions.
- Enhanced diagnostic plotting capabilities for motion correction analysis.

iezqrom committed:
Auto-format code

iezqrom committed:
Add spatial threshold metrics calculation to metrics module

- Introduced `calculate_spatial_threshold_metrics` function for computing spatial threshold metrics on 2D activity images.
- Updated documentation to reflect the new functionality and its parameters.
- Modified __init__.pyi to export the new metrics function.

iezqrom committed:
Auto-format code

iezqrom committed:
Update output filenames in motion plotting functions for consistency

- Changed saved figure filenames from "02_motion_trajectory.png" to "motion_trajectory.png" and from "03_stability_maps.png" to "stability_maps.png".
- Updated corresponding log messages to reflect the new filenames.

iezqrom committed:
Refactor derivative calculations in TraceMetrics for improved accuracy

- Replaced np.diff with np.gradient for calculating first and second derivatives to enhance accuracy.
- Updated variable names for clarity and consistency.
- Adjusted logic for determining window start and end indices based on the new derivative calculations.

iezqrom committed:
Auto-format code

iezqrom committed:
Update gpio dependency in pyproject.toml to conditionally include lgpio for Linux platforms

iezqrom committed:
Add configuration and motion correction features to WidefieldAnalysis

- Introduced DataConfig class to manage file paths for raw and processed trial data.
- Updated WidefieldAnalysis to support both raw and processed data configurations.
- Implemented motion correction pipeline with methods for applying corrections and saving outputs.
- Enhanced data loading methods to accommodate new imaging file formats (TIFF and NPY).
- Added metadata extraction and saving functionalities for trial data.
- Updated tests to cover new features and configurations, ensuring robust functionality.

iezqrom committed:
Auto-format code

iezqrom committed:
Update baseline temperature range in TCSStimulus class documentation and validation

- Changed baseline temperature range from 20-45°C to 16-45°C in the class description and validation constraints.

vkaramanis commented on poulet_py/widefield/io.py:
i think most of these functions are too specific to be into the poulet_py. if in future we change the format of widefield these function will render useless. also some dont do much more than calling directly the function of the library so they are kind of obsolete. the tiff_to_numpy is a good general function which can be added to the tools/serializers.  regarding metadata we can standardize a schema for widefield and then create the necessary functions to parse this. maybe we should create a schemas folder in root of poulet_py for adding all json schemas of different experiments

iezqrom replied:
thanks for the feedback

iezqrom replied:
regarding adding tiff_to_numpy  to tools/serializers

how can i start this process? in this branch? any guidelines

iezqrom replied:
regarding the functions that are just wrappers, you are right, but it'd be beneficial for the user to have one or a few functions to load the data into a script, how should we write this so it's not too specific

iezqrom replied:
regarding the schema, that's a great idea, i got it started for widefield in the last commit, let's discuss this on wednesday

vkaramanis commented on poulet_py/widefield/metrics.py:
i see two option here.
either we create a specific subfolder under tools for movie manipulation and put inside these functions or make them even more general for x n_dim numpy arrays not only 3d.

for movie specific:
i suggest the following structure
```
percentile(x: np.ndarray, percentile: float | list[float] = 15.0,from: int | None = None,to: float | None = None,fps: float | None = None)
```
where  **from** and  **to** is in **ms** and **percentile** can be a list to return multiple percentiles.
similar logic to the other functions

check also [sliding_window_view](https://numpy.org/doc/stable/reference/generated/numpy.lib.stride_tricks.sliding_window_view.html) which might be helpful for windowing and [structured numpy arrays](https://numpy.org/doc/stable/user/basics.rec.html) for the last function which returns a dictionary

iezqrom replied:
> either we create a specific subfolder under tools for movie manipulation and put inside these functions or make them even more general for x n_dim numpy arrays not only 3d.

i'd suggest the first option for now, until we see need in x_dim

I implemented the other suggestions:

- structure
- structured numpy arrays

I'm not sure about sliding window, we're not doing a moving mean. I don't see how to integrate this at the moment

please let me know your thoughts

vkaramanis commented on poulet_py/widefield/motion.py:
Also here some functions are just a encluse of other libraries functions. check if we really need them. if yes we can also add them to the tools/array_manipulation/movie.py or something similar

iezqrom replied:
which ones are a simple wrapper?

I found
_compute_phase_difference
_compute_registration_error
and the ones for saving and loading

what should we do with them?

the other functions are more complex and useful

vkaramanis commented on poulet_py/widefield/analysis.py:
We can make a general Movie class where can include all the manipulations needed the pipeline for widefield analysis most probably should be under the experiment's repo but we can see in more depth when i have the full picture of it.

iezqrom replied:
I agree with making a more general movie class with helper functions for different types of movies but I think this code should go in here. It's for showing videos of WF and we are analysis it right now, the experiment repo will take longer and it'll be general

iezqrom committed:
Update baseline temperature range in TCSStimulus class documentation and validation constraints

iezqrom committed:
Add rolling baseline calculation and update spatial threshold metrics return type

- Added `calculate_rolling_baseline_movie` function to compute rolling baselines for 3D data.
- Updated `calculate_spatial_threshold_metrics` to return a structured numpy scalar instead of a dictionary, enhancing data access and type safety.
- Modified tests to include validation for the new rolling baseline function and updated assertions for the structured output of spatial threshold metrics.

iezqrom committed:
Auto-format code

iezqrom committed:
Refactor calculate_percentile_movie to support time window and multiple percentiles

- Updated `calculate_percentile_movie` to accept a time window defined by `from_ms` and `to_ms`, enhancing flexibility in percentile calculations.
- Modified the `percentile` parameter to accept a list of percentiles, allowing for simultaneous calculations of multiple percentiles.
- Removed references to the deprecated `calculate_rolling_baseline_movie` function in the metrics module and tests.
- Added new tests to validate the functionality of the time window and multiple percentile features.

iezqrom committed:
Merge branch 'spatial-analysis' of https://github.com/poulet-lab/poulet_py into spatial-analysis

iezqrom committed:
Auto-format code

iezqrom committed:
preprocessing

iezqrom committed:
Merge branch 'spatial-analysis' of https://github.com/poulet-lab/poulet_py into spatial-analysis

iezqrom committed:
Auto-format code
